oauth/refresh.go

```diff
@@ -0,0 +1,55 @@
+package oauth
+
+import (
+	"fmt"
+	"net/url"
+
+	"golang.org/x/oauth2"
+)
+
+// RefreshTokenSource will use a refresh token to try and get a new token before
+// calling the original token source to get a new token.
+type RefreshTokenSource struct {
+	// ClientID of the application
+	ClientID string
+
+	// TokenURL is used to fetch new tokens
+	TokenURL string
+
+	// EndpointParams are extra URL query parameters to include in the request
+	EndpointParams *url.Values
+
+	// RefreshToken from a cache, if available. If not, then the first time a
+	// token is requested it will be loaded from the token source and this value
+	// will get updated if it's present in the returned token.
+	RefreshToken string
+
+	// TokenSource to wrap to fetch new tokens if the refresh token is missing or
+	// did not work to get a new token.
+	TokenSource oauth2.TokenSource
+}
+
+// Token generates a new token using either a refresh token or by falling
+// back to the original source.
+func (ts RefreshTokenSource) Token() (*oauth2.Token, error) {
+	if ts.RefreshToken != "" {
+		payload := fmt.Sprintf("grant_type=refresh_token&client_id=%s&refresh_token=%s", ts.ClientID, ts.RefreshToken)
+
+		token, err := requestToken(ts.TokenURL, payload)
+		if err == nil {
+			return token, err
+		}
+
+		// Couldn't refresh, try the original source again.
+	}
+
+	token, err := ts.TokenSource.Token()
```

Comment:
On the next line the check for an error and return it, meaning it will stop and the program will likely display an error and exit. You could intercept the error and retry, but currently we don't do this.

Reply:
Ohhh I completely misinterpreted this, thinking the function was calling itself in the failure case. Thanks for explaining it!